### ISSUE TITLE
Allow conversion between rationals and floats in Dafny runtime

### ISSUE THREAD

**Issue by atomb** [dafny-lang/dafny#1940]
The current Dafny runtime (for C#, at least) uses arbitrary precision rationals to represent real numbers, which provides nice precision but interacts awkwardly with external code that uses floating point. It would be convenient to allow external code to easily convert these values to and from floating-point values.

**Comment by atomb** [comment]
Some additional comments:

* Converting from floats to arbitrary-precision rationals is lossless, so, for example, [this class](https://github.com/dafny-lang/dafny/blob/master/Source/DafnyRuntime/DafnyRuntime.cs#L1506) could have an extra constructor that takes a `double` as an argument without compromising soundness.
* Converting from arbitrary-precision rationals to floats can lose precision, so we would want to name any relevant method in such a way that it's obvious that it's doing an approximation.
* We could also add floating-point types to Dafny, and compile them to floating-point types in target languages. Boogie and Z3 both support reasoning about floating-point values, so it would be a natural fit. I think the change proposed in this issue would still occasionally be valuable in that setting as well, however.

**Comment by MikaelMayer** [comment]
Oh i see, interesting! You'd actually convert:

`b.bbbbbbb * 2^x`

to the big rational

`bbbbbbbb / 2^x`

That makes sense.

**Comment by atomb** [comment]
Yep! And I've worked out the details of how all the calculations would need to happen.

**Comment by RustanLeino** [comment]
A few more points.
* Floating numbers include the value `NaN`, which cannot be converted to a `real`. To work around this, the conversion routine can use some precondition or can return an `Option`.
* To keep verification sound, the routine that returns an approximation of the `real` in a floating number number should be careful not to promise, in a specification, that the two numbers are exactly equal (or that converting back to a `real` gets you the same `real` that you started with).
* If floating point numbers are ever added to Dafny (and I'm not about to suggest this), then they can translate into Boogie's floating point numbers. Boogie already supports them and so does SMT Lib.
